### PR TITLE
KAFKA-5876: IQ should throw different exceptions for different errors(part 1)

### PR DESCRIPTION
**Pull request by vitojeng**
> [KIP-216: IQ should throw different exceptions for different errors](https://cwiki.apache.org/confluence/display/KAFKA/KIP-216%3A+IQ+should+throw+different+exceptions+for+different+errors)

KAFKA-5876's PR break into multiple parts. This PR is part 1: the new exceptions introduced in KIP-216

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
test this please

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java
Comment (vitojeng) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java]: I guess here should not a exception about invalid partition and not sure which type of InvalidStateStoreException we should use. Using InvalidStateStorePartitionException here maybe not a correct choice. 
Any comment?

**Comment by mjsax** [comment]
Thanks for the PR @vitojeng -- I put it into my review backlog :)

**Comment by mjsax** [comment]
@vitojeng This is a quite big PR -- would it make sense to break it into multiple? Maybe one that only add the new exception classed, and one each for using a new exception type?

**Comment by vitojeng** [comment]
> @vitojeng This is a quite big PR -- would it make sense to break it into multiple? Maybe one that only add the new exception classed, and one each for using a new exception type?

@mjsax The same feeling. Break it into multiple part is a good way. I will force-push the new exceptions only in this PR first.

**Comment by mjsax** [comment]
Hey @vitojeng -- sorry for the long delay... We had a big KIP review back log... I would dedicate review time now to get KIP-216 merged. Are you still interested in pushing it over the finish line? If yes, I would start with this PR in the next days.

**Comment by vitojeng** [comment]
> Hey @vitojeng -- sorry for the long delay... We had a big KIP review back log...

@mjsax Never mind, I understand. :)
I'll be rebase master again and update this PR ASAP.

**Comment by vitojeng** [comment]
@mjsax Just rebase & force-pushed.
You can start review when you available.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Jenkins does not cooperator right now. Will try again later.

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (brary) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java:24]: If we are not gonna use this class at all to throw exceptions, is there any need for it? 
Doesn't look like it is solving any extra purpose than StreamsException. Can we directly extend StreamsException in the subclasses which are extending this class?

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java
Comment (brary) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java]: "Indicates that the specific state store being queried via StoreQueryParams used an invalid partition"? 
Does this sound better?

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java
Comment (brary) [inline streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java]: * Indicates that the state store being queried is closed although the Kafka Streams state is
* {@link org.apache.kafka.streams.KafkaStreams.State#RUNNING RUNNING} or
* {@link org.apache.kafka.streams.KafkaStreams.State#REBALANCING REBALANCING}.
* It could happen because the partition moved to some other instance during a rebalance so
* rediscovery of the state store is required before retrying.

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java
Comment (brary) [inline streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java]: *Indicates that the state store being queried is already closed. It could happen when Kafka Streams is in 
* {@link org.apache.kafka.streams.KafkaStreams.State#PENDING_SHUTDOWN PENDING_SHUTDOWN} or
* {@link org.apache.kafka.streams.KafkaStreams.State#NOT_RUNNING NOT_RUNNING} or
* {@link org.apache.kafka.streams.KafkaStreams.State#ERROR ERROR} state.

File: streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java
Comment (brary) [inline streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java]: Indicates that the state store being queried is invalid. Hence, it will be futile to retry again.

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java:24]: @brary Thanks for reviewing. In KIP-216, the main purpose of keeping InvalidStateStoreException is to keep compatibility, this should be important to users. Your suggestion still valuable, we can do it in the future.

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java]: Looks great! 
BTW, I would use StoreQueryParameters instead of StoreQueryParams. :)

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java]: Copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java]: Copy that.

File: streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java]: Copy that.

**Comment by vitojeng** [comment]
Address comments and rebase trunk.

**Comment by vitojeng** [comment]
Retest this please.

**Comment by mjsax** [comment]
@vitojeng -- Jenkins got locked down a couple of week ago. Only committer can trigger builds now.

**Comment by vitojeng** [comment]
Got it!

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java:24]: Sure @vitojeng . cc @mjsax

**Comment by brary** [comment]
LGTM!

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java]: nit: JavaDocs does support HTML, but `<p>` is actually not HTML but a JavaDoc annotation and the closing tag `</p>` is not required. In fact, we only have a single paragraph and thus don't need any tags at all.

`InvalidStateStoreException` -> `{@code InvalidStateStoreException}`

`...but only it subclasses.`

I would not list the sub-classes because the JavaDocs will link to them automatically. Also, if we add new sub-classes in the future, I am sure we would forget to update the JavaDocs here. If we don' list them, we can't forget to update as there is nothing to be updated. :)

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java]: It's a little bit unclear what "invalid" means. Maybe better:
```
...used a partitioning that is not assigned to this instance. You can use {@link KafkaStreams#allMetadata()} to discover the correct instance that hosts the requested partition.
```

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java]: `It` -> `This`

`partition` -> `store` ?

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java]: `To query state stores, it's required to first start Kafka Streams via {@link KafkaStreams#start()}. You can retry to query the state after the state transitioned to ...`

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java]: `It` -> `This`

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java]: `Indicate[s] that Kafka Streams is in state {@link org.apache.kafka.streams.KafkaStreams.State#CREATED CREATED} and thus state stores cannot be queries yet.`

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsRebalancingException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StreamsRebalancingException.java]: `Indicate[s] that Kafka Streams is in state {@link org.apache.kafka.streams.KafkaStreams.State#REBALANCING REBALANCING} and thus cannot be queried by default.
You can retry to query after the rebalance finished. As an alternative, you can also query (potentially stale) state stores during a rebalance via {@link StoreQueryParameters#enableStaleStores()}.`

File: streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java]: `invalid` -> `unknown, i.e., the state store does either not exist in your topology or it is not queryable.`

I would remove `Hence, it will be futile to retry again.`

**Review by mjsax — COMMENTED** [review]
Sorry (again... :() for the delayed review. Couple of minor suggestions to improve the JavaDocs. The goal of the KIP was really to make it easier for users to reason about the type of error and thus we should be a little bit more elaborative.

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java]: Got it!

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StateStoreMigratedException.java]: Nice catch. It's my bad... :(

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java:22]: Got it.

**Comment by vitojeng** [comment]
Sorry, should not rebase...
I'll force-push again!

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StateStoreNotAvailableException.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsRebalancingException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/StreamsRebalancingException.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/UnknownStateStoreException.java]: copy that

**Comment by vitojeng** [comment]
@mjsax Thanks for the review. Already address your comment. :)

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java
Comment (vitojeng) [reply streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStorePartitionException.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java]: Nit: we don't need the `<p>` tag because there is only one paragraph.

File: streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/InvalidStateStoreException.java]: Nit: `Note that {@code InvalidStateStoreException} [is] not thrown directly but only [its] sub-classes.`

File: streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/StreamsNotStartedException.java]: typo `to to`

(also missing `.` at the end of the sentence)

**Review by mjsax — COMMENTED** [review]
Thanks for the update @vitojeng -- a few more nits. Overall LGTM. Will merge after addressed.

**Comment by vitojeng** [comment]
> Thanks for the update @vitojeng -- a few more nits. Overall LGTM. Will merge after addressed.

Thanks @mjsax for the reviewing, just updated.

**Comment by mjsax** [comment]
Retest this please.

**Review by mjsax — APPROVED** [review]
LGTM.

Will merge after Jenkins passed.

**Comment by mjsax** [comment]
Merged to `trunk`. Thanks for the PR and patience @vitojeng!

Looking forward to the next PR(s) -- it's best to continue doing multiple smaller PRs (I leave it up to your judgment what the best way is to split the work in chunks.)

**Comment by vitojeng** [comment]
> Looking forward to the next PR(s) -- it's best to continue doing multiple smaller PRs (I leave it up to your judgment what the best way is to split the work in chunks.)

Thanks @mjsax !
I would separate each new exception into different PRs. I guess this should be a better way for reviewing. Also thanks @brary for the reviewing.

**Pull request merged**
